core/includes/file.inc

```diff
@@ -1496,208 +1496,233 @@ function file_space_used($uid = NULL, $status = FILE_STATUS_PERMANENT) {
- *   Related topics, or the header at the top of this file, documents the
- *   components of a file entity. In addition to the standard components,
- *   this function adds:
+ * @param int $delta
+ *   If multiple files are uploaded into the same field (via the "multiple"
+ *   HTML attribute), an individual file can be saved one at a time by
+ *   specifying numeric offset for that file, starting from 0.
+ *
+ * @return File|File[]|FALSE|NULL
+ *   A File object (or array of File objects) containing the file information if
+ *   the upload succeeded, FALSE in the event of an error, or NULL if no file
+ *   was uploaded.
+ *
+ *   If only a single file was uploaded, or $delta was specified, then a single
+ *   File object is returned. If multiple files are uploaded, an array of
+ *   return values is provided instead.
+ *
+ *   The documentation for the "File interface" group, which you
+ *   can find under Related topics, or the header at the top of this file,
+ *   documents the properties of a File object. After saving, each File object
+ *   will be populated with the following additional properties:
  *   - source: Path to the file before it is moved.
  *   - destination: Path to the file after it is moved (same as 'uri').
  */
-function file_save_upload($form_field_name, $validators = array(), $destination = FALSE, $replace = FILE_EXISTS_RENAME) {
+function file_save_upload($form_field_name, $validators = array(), $destination = FALSE, $replace = FILE_EXISTS_RENAME, $delta = NULL) {
   global $user;
   static $upload_cache;
 
-  // Return cached objects without processing since the file will have
-  // already been processed and the paths in _FILES will be invalid.
-  if (isset($upload_cache[$form_field_name])) {
-    return $upload_cache[$form_field_name];
-  }
-
   // Make sure there's an upload to process.
   if (empty($_FILES['files']['name'][$form_field_name])) {
     return NULL;
   }
 
-  // Check for file upload errors and return FALSE if a lower level system
-  // error occurred. For a complete list of errors:
-  // See http://php.net/manual/features.file-upload.errors.php.
-  switch ($_FILES['files']['error'][$form_field_name]) {
-    case UPLOAD_ERR_INI_SIZE:
-    case UPLOAD_ERR_FORM_SIZE:
-      backdrop_set_message(t('The file %file could not be saved because it exceeds %maxsize, the maximum allowed size for uploads.', array('%file' => $_FILES['files']['name'][$form_field_name], '%maxsize' => format_size(file_upload_max_size()))), 'error');
-      return FALSE;
-
-    case UPLOAD_ERR_PARTIAL:
-    case UPLOAD_ERR_NO_FILE:
-      backdrop_set_message(t('The file %file could not be saved because the upload did not complete.', array('%file' => $_FILES['files']['name'][$form_field_name])), 'error');
-      return FALSE;
+  // Check if this is a multiple file upload. If there is only a single file,
+  // return the first and only file (delta 0).
+  $multiple_file_upload = is_array($_FILES['files']['name'][$form_field_name]);
```

Comment:
Wondering if we should always be returning an array of File objects, and have this here be a `count()` instead. So cases would be:
- array with a single item
- array with multiple items
- empty array (nothing uploaded)

That would simplify the type of the return of this function to array|FALSE.